hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java

```diff
@@ -202,12 +206,40 @@ public String writeIds(String name, Collection<Id> ids) {
 
     @Override
     public String writePaths(String name, Collection<HugeTraverser.Path> paths,
-                             boolean withCrossPoint) {
+                             boolean withCrossPoint,
+                             Iterator<Vertex> vertices) {
         List<Map<String, Object>> pathList = new ArrayList<>(paths.size());
         for (HugeTraverser.Path path : paths) {
             pathList.add(path.toMap(withCrossPoint));
         }
-        return writeList(name, pathList);
+
+        Map<String, Object> results;
+        if (vertices == null) {
+            results = ImmutableMap.of(name, pathList);
+        } else {
+            results = ImmutableMap.of(name, pathList, "vertices", vertices);
+        }
+        return writeObject(results);
+    }
+
+    @Override
+    public String writeCrosspoints(
+                  CustomizedCrosspointsTraverser.CrosspointsPaths paths,
+                  Iterator<Vertex> iterator, boolean withPath) {
+        Map<String, Object> results;
+        List<Map<String, Object>> pathList;
+        if (withPath) {
+            pathList = new ArrayList<>();
+            for (HugeTraverser.Path path : paths.paths()) {
+                pathList.add(path.toMap(false));
+            }
+        } else {
+            pathList = ImmutableList.of();
+        }
+        results = ImmutableMap.of("crosspoints", paths.crosspoints(),
+                                  "paths", pathList,
+                                  "vertices", iterator);
```

Comment:
may iterator be null?

Reply:
won't be null.
Collections.emptyIterator()
or 
empty Metadatable Iterator